src/Quaternion.jl

```diff
@@ -47,17 +47,47 @@ julia> quat([1, 2, 3])
  Quaternion{Int64}(3, 0, 0, 0)
 ```
 """
-quat
+function quat end
 
-quat(p, v1, v2, v3) = Quaternion(p, v1, v2, v3)
-quat(x) = Quaternion(x)
+quat(q::Quaternion) = q
+quat(s::Real) = Quaternion(s)
+quat(s::Real, v1::Real, v2::Real, v3::Real) = Quaternion(s, v1, v2, v3)
+
+## Array operations on quaternions ##
+quat(A::AbstractArray{<:Quaternion}) = A
 function quat(A::AbstractArray{T}) where T
     if !isconcretetype(T)
         error("`quat` not defined on abstractly-typed arrays; please convert to a more specific type")
     end
     convert(AbstractArray{typeof(quat(zero(T)))}, A)
 end
 
+"""
+    quat(T::Type)
+
+Return an appropriate type which can represent a value of type `T` as a quaternion.
```

Comment:
```suggestion
Return an appropriate type that can represent a value of type `T` as a quaternion.
```

Reply:
The docstring for `complex(::Type)` uses "which", but it seems "that" sounds more natural, so I will update as suggested.

https://github.com/JuliaLang/julia/blob/e7345b89fd4eb15e8f395395701e19be705d7b06/base/complex.jl#L180

Reply:
Yeah, "which" would convey a different in incorrect meaning here (also in Base).